predicators/envs/ball_and_cup_sticky_table.py

```diff
@@ -0,0 +1,523 @@
+        cup_geom = self._object_to_geom(cup, state)
+        assert isinstance(ball_geom, utils.Circle)
+        assert isinstance(cup_geom, utils.Circle)
+        ball_and_cup_at_same_pos = cup_geom.contains_circle(ball_geom)
+        holding_ball = self._Holding_holds(state, [ball])
+        holding_cup = self._Holding_holds(state, [cup])
+        return ball_and_cup_at_same_pos and (
+            (holding_ball and holding_cup) or
+            (not holding_ball and not holding_cup))
+
+    def _BallNotInCup_holds(self, state: State,
+                            objects: Sequence[Object]) -> bool:
+        return not self._BallInCup_holds(state, objects)
+
+    def _table_is_sticky(self, table: Object, state: State) -> bool:
+        return state.get(table, "sticky") > 0.5
+
+    def simulate(self, state: State, action: Action) -> State:
+        # NOTE: noise is added here. Two calls to simulate with the same
+        # inputs may produce different outputs!
+        assert self.action_space.contains(action.arr)
+        move_or_pickplace, obj_type_id, ball_only, act_x, act_y = action.arr
+        next_state = state.copy()
+        hand_empty = self._HandEmpty_holds(state, [])
+        ball, = state.get_objects(self._ball_type)
+        cup, = state.get_objects(self._cup_type)
+        robot, = state.get_objects(self._robot_type)
+        ball_held = self._Holding_holds(state, [ball])
+        cup_held = self._Holding_holds(state, [cup])
+        ball_in_cup = self._BallInCup_holds(state, [ball, cup])
+        obj_being_held: Optional[Object] = None
+        if (ball_held and not cup_held) or ball_only > 0.5:
+            obj_being_held = ball
+        elif cup_held:
+            obj_being_held = cup
+        # In this case, handle picking/placing.
+        if move_or_pickplace == 1.0:
+            # Picking logic.
+            if hand_empty:
+                # Fail sometimes.
+                if self._noise_rng.uniform() < self._pick_success_prob:
+                    if obj_type_id == 1.0:
+                        # Pick ball.
+                        if self._action_grasps_object(act_x, act_y, ball,
+                                                      state):
+                            next_state.set(ball, "held", 1.0)
+                        assert self._Holding_holds(next_state, [ball])
+                    else:
+                        assert obj_type_id == 2.0
+                        if self._action_grasps_object(act_x, act_y, cup,
+                                                      state):
+                            # Pick cup.
+                            next_state.set(cup, "held", 1.0)
+                            if ball_in_cup:
+                                # Pick both ball and cup simultaneously.
+                                next_state.set(ball, "held", 1.0)
+                                assert self._Holding_holds(next_state, [ball])
+                        assert self._Holding_holds(next_state, [cup])
```

Comment:
this should be indented once more right? I guess it hasn't crashed because ```self._action_grasps_object(act_x, act_y, cup, state)``` always succeeds?

Reply:
yes - great catch!